docs/source/dataframe-design.rst

```diff
@@ -0,0 +1,109 @@
+Internal Design
+===============
+
+Dask dataframes coordinate many Pandas arrays arranged along an index. We
+define a ``dask.dataframe`` object with the following components:
+
+- A dask graph with a special set of keys designating partitions, such as
+  ``('x', 0), ('x', 1), ...``.
+- A name to identify which keys in the dask graph refer to this dataframe, such
+  as ``'x'``.
+- An empty pandas object containing appropriate metadata (e.g.  column names,
+  dtypes, etc...).
+- A sequence of partition boundaries along the index.
+
+Metadata
+--------
+
+Many dataframe operations rely on knowing the name and dtype of columns. To
+keep track of this information, all ``dask.dataframe`` objects have a ``_meta``
+attribute which contains an empty object with the same dtypes and names. For
+example:
+
+.. code-block:: python
+
+   >>> df = pd.DataFrame({'a': [1, 2, 3], 'b': ['x', 'y', 'z']})
+   >>> ddf = dd.from_pandas(df, npartitions=2)
+   >>> ddf._meta
+   Empty DataFrame
+   Columns: [a, b]
+   Index: []
+   >>> ddf._meta.dtypes
+   a     int64
+   b    object
+   dtype: object
+
+Internally dask does its best to propogate this information through all
+operations, so most of the time a user shouldn't have to worry about this.
+Usually this is done by evaluating the operation on a small sample of fake
+data, which can be found on the ``_meta_nonempty`` attribute:
+
+.. code-block:: python
+
+   >>> ddf._meta_nonempty
+      a    b
+   0  1  foo
+   1  1  foo
+
+Sometimes though this operation may fail, or may be prohibitively expensive.
```

Comment:
That falls under the "may fail" case, but I'm happy to provide an example of why it might fail.